Allow adding error symbols in support modules

EmilioPeJu opened the pull request:
This was acomplished by making errSymbolAdd add the error symbol directly into the global hash table and removing errnumlist which is not needed anymore.

Unit tests were added for checking the following cases:
- Adding a valid symbol and checking that it exists (fixed by this change)
- Getting an existing error symbol
- Getting a non existing error symbol
- Adding an invalid error symbol (fixed by this change)
- Adding an error symbol with a code that already exists (fixed by this change)

Therefore, issue #268 was fixed

EmilioPeJu committed:
Allow adding error symbols after early initialization

This was acomplished by making errSymbolAdd add the error symbol directly
into the global hash table and removing errnumlist which is not needed
anymore.

Unit tests were added for checking the following cases:
- Adding a valid symbol and checking that it exists (fixed by this change)
- Getting an existing error symbol
- Getting a non existing error symbol
- Adding an invalid error symbol (fixed by this change)
- Adding an error symbol with a code that already
  exists (fixed by this change)

Therefore, issue #268 was fixed

EmilioPeJu commented:
should we pull the constant `501` to a define?  it appears a few times

mdavidsaver commented:
@anjohnson If we are going to encourage custom error codes in user code, should we take some steps to prevent collisions?

EmilioPeJu commented:
BTW, I also did the change for not allowing double adding the same error code (not sure if that's what you meant).

One thing to consider is how to make it easier to add error symbols in a support modules (an easier way other than having to call `errSymbolAdd` in some driver init function), it could be something similar to what epics base does, extracting the symbols from from some specified headers files

EmilioPeJu committed:
error: statically allocate error symbol hash table

This will allow calling errSymbolAdd before errSymBld, therefore, a
function adding error symbols can now be run before iocInit or errlogInit

EmilioPeJu commented:
I realised that the size of the error symbol hash table is known at compile time, so, I changed it to be allocated statically (having the extra advantage of allowing adding error symbols much earlier). Please, note that because it's in the bss section, it will be zero which is the proper initial state for the hash table

EmilioPeJu committed:
error: add a constant for the minimum module number

anjohnson commented:
Could you modify the check for duplicate symbols being added so it doesn't return an error if both the number and the name string are identical to an already-registered entry? This reduces unnecessary errors in circumstances like testing where the symbol registration code might be run more than once — this should be allowed.

#265 says:
> there are currently no locks protecting the data structures in errSymLib.c so they need adding as well, preferably in a way that doesn't require a separate lock/unlock for each symbol that gets added.

I'm not seeing any added `epicsMutex` objects in your changes to protect the hashtable from being modified while it is being searched, was this something you looked at?

The other thing I would like to see is an entry in the `documentation/RELEASE_NOTES.md` file explaining how external software can add its private error symbols to the table.

EmilioPeJu committed:
Make adding an identical error symbol not fail

A test case was also added which test that adding an error symbol
with same error code and message as one added before will not fail

EmilioPeJu committed:
Add locking to error symbol table

This protects the cases of:
- simultaneously adding and requesting of an error symbol
- simultaneously adding many error symbols

x

EmilioPeJu committed:
Update release notes regarding adding error symbols

EmilioPeJu commented:
Done, instead of using an `epicsMutex`, the add operation is done atomically with `epicsAtomicSetPtrT`, each position of the error symbol hash table keeps a single linked list, therefore, adding just requires to set a pointer.

mdavidsaver commented:
Atomic container algorithms are tricky to get correct...

fyi. I think that you would need to use a loop with `epicsAtomicCmpAndSwapPtrT()` instead of `epicsAtomicSetPtrT()` to prevent concurrent calls (however unlikely) from leaking memory.  Also, it is generally advisable to use atomic operations for all access to a variable.  `epicsAtomicGetPtrT()` instead of simple dereference.

exzombie commented on modules/libcom/src/error/errSymLib.c:
```diff
@@ -23,25 +23,25 @@
 #include "ellLib.h"
 #include "errlog.h"
 
+#define MIN_MODULE_NUM 501
```
I would put this into `errMdef.h`, next to the existing comment explaining the reasoning behind the number.

EmilioPeJu replied:
Will do, thanks for the suggestion

EmilioPeJu commented:
I'm not protecting the case of simultaneous adding symbols on purpose because I don't think the extra complexity is worth it. The error symbols will be added in some init function (for example driver init), which are called in the main thread and therefore, there is not such risk.

As I said earlier, the focus was in the case of simultaneously adding and fetching a symbol, which is something that could happen because any error logging  could fetch a symbol (which is something that could easily happen in a thread parallel to the main thread adding)

exzombie commented on modules/libcom/src/error/errSymLib.c:
```diff
@@ -23,25 +23,25 @@
 
-static ELLLIST errnumlist = ELLLIST_INIT;
-static ERRNUMNODE **hashtable;
+static ERRNUMNODE *hashtable[NHASH * sizeof(ERRNUMNODE *)];
```
Inappropriate use of `sizeof`: this a static array, not `malloc()`, so its size is given as number of elements, not bytes.

EmilioPeJu replied:
Please note that we are allocating pointers, so, what we have is the same as `sizeof(void *)`

exzombie replied:
Let me put this another way. The following three statements allocate an object of the same size:
```
void *dynmem = malloc(num_elements * sizeof(void*));
void *zeromem = calloc(num_elements, sizeof(void*));
void *staticmem[num_elements];
```

EmilioPeJu replied:
Ohh, now I understand, thanks for spotting that, I'll fix it

exzombie commented on modules/libcom/src/error/errSymLib.c:
```diff
@@ -23,25 +23,25 @@
 static epicsUInt16 errhash(long errNum);
 
 typedef struct errnumnode {
-    ELLNODE            node;
     long               errNum;
     struct errnumnode *hashnode;
     const char        *message;
     long               pad;
 } ERRNUMNODE;
```
As I understand it up to this point, the removed `node` field and the `hashnode` field have overlapping roles: both make this struct an element of a linked list. Personally, I don't mind a "hand-rolled" linked list, but in that case, the `hashnode` field should be renamed to make it obvious what it represents. On the other hand, it would be preferable to use a consistent API for linked lists everywhere, which means removing the `hashnode` field and restoring `node` instead.

Edit: ok, I get it now. This thing used to be a part of two linked lists, and if you keep `hashnode` instead of `node`, you get to keep existing code for iterating over the hashmap list. I would use this opportunity to change the code, but ultimately, it's up to you. It's just style, after all.

EmilioPeJu replied:
The linked list we had before was used only as an intermediate list that is iterated and added to the hash table, so I optimized that and now the `errSymbolAdd` function adds directly to the hash table. The linked list inside the hash table doesn't require a linking to the previous node, therefore the current approach is simpler and with less overhead than using `ELLNODE`

exzombie commented on modules/libcom/src/error/errSymLib.c:
```diff
@@ -56,44 +56,17 @@ extern ERRSYMTAB_ID errSymTbl;
 int errSymBld(void)
```
The description of the function can be simplified, just as the function itself has been.

EmilioPeJu replied:
Done

exzombie commented:
> As I said earlier, the focus was in the case of simultaneously adding and fetching a symbol, which is something that could happen because any error logging could fetch a symbol (which is something that could easily happen in a thread parallel to the main thread adding)

If you choose to limit the support to that use case, that's sufficient reasoning for using atomic-set instead of compare-and-swap, but you should explain it in a comment somewhere so that it's clear for future readers that writing from multiple threads is not supported. However, Michael's point that you need to use atomic-get to read the values stands: if you don't, there is no guarantee that long-lived thread would *ever* get an updated value.

Edit: thinking about it again, it's probably easier to just do compare-and-swap than try to explain the reasoning in a comment :)

exzombie commented on modules/libcom/src/error/errSymLib.c:
```diff
@@ -109,21 +82,47 @@ static epicsUInt16 errhash(long errNum)
 
     modnum = (unsigned short) (errNum >> 16);
     errnum = (unsigned short) (errNum & 0xffff);
```
This is a very common pattern here, so I would put it in a function, even if it is just two lines of code.

EmilioPeJu replied:
what do you think about having 2 macros at the top of the file: `ERROR_NUMBER(status)` and `ERROR_MODULE(status)`?

mdavidsaver commented:
> ... I'm not protecting the case of simultaneous adding symbols on purpose because I don't think the extra complexity is worth it. ...

I'm don't think intentionally leaky code is acceptable.  The simple way would be to use a mutex and live with a likely minor performance hit when looking up error messages (hopefully a relatively rare occurrence).

EmilioPeJu commented:
Thanks @exzombie and @mdavidsaver for reviewing, I have added the epics mutex as you suggested, I hope I got it right this time

mdavidsaver commented on documentation/RELEASE_NOTES.md:
```diff
@@ -17,6 +17,14 @@ should also be read to understand what has changed since earlier releases.
 
+### Support modules can now add error symbols even before calling `errlogInit`
+
+A support module can add an error symbol by calling `errSymbolAdd`, for
+example: `errSymbolAdd(S_newError, "New error message")`
+
+This used to work in older EPICS versions but got broken at some point and is
+now restored.
+
```
I would describe this change as "Allow errSymbolAdd() to be called after early initialization".  I would also like to see some guidance to users about how avoid error number collisions.  Also, "got broken at some point" is not helpful to users.

For that matter, I think this sort of information should be added as a doxygen comment for `errSymbolAdd()`.  eg.

```c
/** Add custom error message to be returned by subsequent `errSym*` calls.
 * \since 7.0.2 Calls after errlogInit() have no effect.
 * \since UNRELEASED Can be called at any time.
 */
```

The `7.0.2` is a guess that 0fdb995a0bda9a7d9591b6c2b1bde5b2aab43421 contained the "breakage".

EmilioPeJu replied:
I simplified the release note using the text you suggest, unfortunately I can't build your guess commit to verify it, not sure if because build was broken or because there is something nasty in my environment

mdavidsaver commented on modules/libcom/src/error/errSymLib.c:
```diff
@@ -33,69 +35,55 @@
+
 extern ERRSYMTAB_ID errSymTbl;
 
+static void populateEpicsErrorSymbols()
+{
+    ERRSYMBOL *errArray = errSymTbl->symbols;
+    int i;
+
+    for (i = 0; i < errSymTbl->nsymbols; i++, errArray++) {
+        if (errSymbolAdd(errArray->errNum, errArray->name)) {
+            fprintf(stderr, "errSymBld: ERROR - errSymbolAdd() failed \n");
```
This new `populateEpicsErrorSymbols()` is only called from `errSymBld()`, which does nothing but call `populateEpicsErrorSymbols()`.  Why introduce another function name?

EmilioPeJu replied:
it was just for having a clearer code (easier to read without needing to go deeper), the new function has a name that describes what it does... however, I'm happy to move it to `errSymBld()` if that is preferred.

EmilioPeJu replied:
OK, I've moved it, given that the big comment at the top says what it does, it was redundant

anjohnson commented:
Core Group: Code now looks good, but needs test results from Appveyor before pushing the merged version.

mdavidsaver committed:
Fix compile w/ vs2012

mdavidsaver commented:
I have rebased this PR on top of a change to fix the appveyor builds.  I have also added an `#include <epicsTypes.h>` which will hopefully fix a build failure with vs2012.

I am removing Approval until an appveyor run passes.

mdavidsaver commented on modules/libcom/src/error/errSymLib.c:
```diff
@@ -109,21 +92,48 @@ static epicsUInt16 errhash(long errNum)
+
+    initErrorHashTable();
+
+    epicsUInt16 hashInd = errhash(errNum);
```
The appveyor vs2012 build is still failing on this line.  My attempt to include `epicsTypes.h` above did not help, which confuses me.  I am not certain why this does not include the typedef for `epicsUInt16`?

```
cl               -nologo -FC -D__STDC__=0 -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -DBUILDING_libCom_API   -Ox -GL -Oy-   -W3        -MD -DEPICS_BUILD_DLL -DEPICS_CALL_DLL    -I. -I../O.Common -I. -I../osi/compiler/msvc -I../osi/compiler/default -I. -I../osi/os/WIN32 -I../osi/os/default -I.. -I../as -I../bucketLib -I../calc -I../cvtFast -I../cppStd -I../cxxTemplates -I../dbmf -I../ellLib -I../env -I../error -I../fdmgr -I../flex -I../freeList -I../gpHash -I../iocsh -I../log -I../macLib -I../misc -I../osi -I../pool -I../ring -I../taskwd -I../timer -I../yacc -I../yacc -I../yajl -I../../../../include/compiler/msvc -I../../../../include/os/WIN32 -I../../../../include         -c ../error/errSymLib.c
errSymLib.c
c:\projects\epics-base\modules\libcom\src\error\errsymlib.c(114) : error C2275: 'epicsUInt16' : illegal use of this type as an expression
        c:\projects\epics-base\modules\libcom\src\misc\epicstypes.h(47) : see declaration of 'epicsUInt16'
c:\projects\epics-base\modules\libcom\src\error\errsymlib.c(114) : error C2146: syntax error : missing ';' before identifier 'hashInd'
c:\projects\epics-base\modules\libcom\src\error\errsymlib.c(114) : error C2065: 'hashInd' : undeclared identifier
c:\projects\epics-base\modules\libcom\src\error\errsymlib.c(117) : error C2065: 'hashInd' : undeclared identifier
make[4]: *** [../../../../configure/RULES_BUILD:258: errSymLib.obj] Error 2
make[4]: Leaving directory 'C:/projects/epics-base/modules/libcom/src/O.win32-x86'
make[4]: *** Waiting for unfinished jobs....
make[4]: Entering directory 'C:/projects/epics-base/modules/libcom/src/O.win32-x86'
cl               -nologo -FC -D__STDC__=0 -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -DBUILDING_libCom_API   -Ox -GL -Oy-   -W3        -MD -DEPICS_BUILD_DLL -DEPICS_CALL_DLL    -I. -I../O.Common -I. -I../osi/compiler/msvc -I../osi/compiler/default -I. -I../osi/os/WIN32 -I../osi/os/default -I.. -I../as -I../bucketLib -I../calc -I../cvtFast -I../cppStd -I../cxxTemplates -I../dbmf -I../ellLib -I../env -I../error -I../fdmgr -I../flex -I../freeList -I../gpHash -I../iocsh -I../log -I../macLib -I../misc -I../osi -I../pool -I../ring -I../taskwd -I../timer -I../yacc -I../yacc -I../yajl -I../../../../include/compiler/msvc -I../../../../include/os/WIN32 -I../../../../include         -c errSymTbl.c
errSymTbl.c
```

https://ci.appveyor.com/project/epics-base-7/epics-base/builds/47895623/job/6a76h0mefu03jsr7#L1429

ralphlange commented:
Maybe it's

> When you name your source files *.c, MSVC assumes it's compiling C, which means C89. All block-local variables need to be declared at the beginning of the block.

From: https://stackoverflow.com/questions/9903582/error-c2275-illegal-use-of-this-type-as-an-expression

mdavidsaver commented:
Ah, yes you are right of course.  I took the error message too literally.

I have tried to push up another change.  Hopefully github.com will pick this up whenever their various issues are resolved.

mdavidsaver commented:
Ok.  that did the trick.

mdavidsaver commented:
Merged as of 88ea1507f46daabb0520d773193b271ac6c9b56d
